### PR TITLE
HIVE-26345: SQLOperation class output real exception message to jdbc client

### PR DESCRIPTION
**Pull request by zhangbutao**
### Why are the changes needed?

See details from  https://issues.apache.org/jira/browse/HIVE-26345
SQLOperation class output real exception message to jdbc client and user can amend the query based on real exception.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Local cluster test.

Before the PR, beeline dispaly exception:
`Error: Error running query (state=,code=0) `

With the PR, beeline dispaly exception:
`Error: Error running query: java.lang.RuntimeException: To use DbTxnManager you must set hive.support.concurrency=true (state=,code=0)`

### CODE REVIEW

**Comment by zhangbutao** [comment]
@belugabehr @pvary @zabetak wdyt? Thank you.

**Review by zabetak — COMMENTED** [review]
Thanks for working on this @zhangbutao and pointing out the behavior changes! 

As I wrote under https://github.com/apache/hive/pull/3364 this message concatenation pattern has trade-offs (between user and developper).

Have you considered attacking this from the beeline side?

**Comment by zhangbutao** [comment]
> 
> As I wrote under #3364 this message concatenation pattern has trade-offs (between user and developper).
> 
> Have you considered attacking this from the beeline side?

@zabetak Thank you for your detailed explanation! I will do some further research based on your suggestion.

**Comment by zhangbutao** [comment]
@zabetak I checked beeline code adn hs2 code, and i still think this change should be fixed in hs2 side. Here's my test, please correct me if i am wrong. Thank you.

1. Use following query to test exception message:
`set hive.support.concurrency=false;`
`set hive.txn.manager=org.apache.hadoop.hive.ql.lockmgr.DbTxnManager;`
`create table testacid(id int) stored as orc tblproperties('transactional'='true');`

2. The final result which wraps exception returned from **HS2 side**, a _**TExecuteStatementResp**_ object
https://github.com/apache/hive/blob/8f1a5b6854daf5fb4814632f9356ef5fe7bb6964/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java#L672-L679
And the final exception massage is set  using _tStatus.setErrorMessage(getMessage());_, that is say client can only get simplified exception message(**_Error running query_**) instead of full exception stacktrace:
https://github.com/apache/hive/blob/8f1a5b6854daf5fb4814632f9356ef5fe7bb6964/service/src/java/org/apache/hive/service/cli/HiveSQLException.java#L131

3. The excepton displayed from the **Beeline side**:
Beeline client gets query response (**_TExecuteStatementResp_** object) from HS2 side, and uses _Utils.verifySuccessWithInfo(execResp.getStatus())_ to check status or throw exception:
https://github.com/apache/hive/blob/8f1a5b6854daf5fb4814632f9356ef5fe7bb6964/jdbc/src/java/org/apache/hive/jdbc/HiveStatement.java#L353-L354
https://github.com/apache/hive/blob/8f1a5b6854daf5fb4814632f9356ef5fe7bb6964/jdbc/src/java/org/apache/hive/jdbc/Utils.java#L378-L383
The final exception message displayed to client is _status.getErrorMessage()_, that is "**_Error running query_**" returned from HS2 side:
https://github.com/apache/hive/blob/8f1a5b6854daf5fb4814632f9356ef5fe7bb6964/service/src/java/org/apache/hive/service/cli/HiveSQLException.java#L117-L118


So, i think the root cause was that HS2 side did not return full valid exception message, and because we often use **_e.getMessage()_** to retrieve meaningful exception, we should add message content when throw a exception, e.g.:
`throw new HiveSQLException("Error running query: " + e.toString(), e);`

**Comment by zabetak** [comment]
Thanks for your detailed analysis @zhangbutao and apologies for the delay. I guess you are right that we should handle the problem in the HS2 side but I still believe that concatenating exception messages is not the right way to move forward.

If we need to inform the client (e.g., beeline) about a specific error maybe we need to use an entry from `ErrorMsg` class, add a new one, and or do something similar to what happens in [Compiler.handleException](https://github.com/apache/hive/blob/714c260e4a7c6b147c897718a33e693699267792/ql/src/java/org/apache/hadoop/hive/ql/Compiler.java#L459).

**Comment by zhangbutao** [comment]
@zabetak After further investigation, i found HIVE-22526 separateed txn related codes from compile codes which lead to exception(errorCode 10264) from txn not being catched and converted into `CommandProcessorException`. I fixed this regression by catch txn exception in txn code block. 

After fixing, we can get txn exception whth errorCode(10264) and state(42000).
`Error: Error while compiling statement: FAILED: RuntimeException [Error 10264]: To use DbTxnManager you must set hive.support.concurrency=true (state=42000,code=10264)`

BTW, for other exception which may not exist in` ErrorMsg class`,  shoud we add a generic error code in `ErrorMsg class` or simply throw message? I have seen lots of code which handle `HiveSQLException` like this:
`throw new HiveSQLException("Error running query: " + e.toString(), e);`

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 1]
```diff
@@ -511,13 +511,17 @@ public void compile(String command, boolean resetTaskIds, boolean deferClose) th
   }
 
   private void prepareForCompile(boolean resetTaskIds) throws CommandProcessorException {
-    driverTxnHandler.createTxnManager();
-    DriverState.setDriverState(driverState);
-    prepareContext();
-    setQueryId();
+    try {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:514]: I suppose that this try-catch block was introduced to address the [so-called regression](https://github.com/apache/hive/pull/3393#issuecomment-1207815099) from HIVE-22526 but I don't fully understand where the regression comes from.

I had a quick look in the diff introduced by HIVE-22526 but as far as I can see there the`driverTxnHandler#createTxnManager` is not wrapped in a try-catch block and it seems that the exception is not passed in `handleException` method. @zhangbutao can you explain a bit more on where exactly is the regression?

Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:514]: @zabetak Maybe I didn't describe it exactly. In fact, I tried to fix two related issues in this PR.

1. First: Beeline client can not get real exception message which is caused by the change(https://issues.apache.org/jira/browse/HIVE-23124).

2. Second: Even though the first issus is fixed, Beeline client can not get valid exception state and code in the following test:
`set hive.support.concurrency=false;`
`set hive.txn.manager=org.apache.hadoop.hive.ql.lockmgr.DbTxnManager;`
`create table testacid(id int) stored as orc tblproperties('transactional'='true');`
_**Error: Error running query: java.lang.RuntimeException: To use DbTxnManager you must set hive.support.concurrency=true (state=,code=0)**_
However, In hive3, we can get valiad exception with state=42000, code=10264
**_Error: Error while compiling statement: FAILED: RuntimeException [Error 10264]: To use DbTxnManager you must set hive.support.concurrency=true (state=42000,code=10264)_**
This change was introduced by https://issues.apache.org/jira/browse/HIVE-22526, which missed handling the exception from `driverTxnHandler#createTxnManager`, i also think we should fix this.

**Pull request closed without merging**